### PR TITLE
keyvisualizer, kvserver: Interfaces and RPCs powering the Key Visualizer

### PR DESCRIPTION
**Pull request by zachlite**
Some skeleton implementation and plumbing is included. This was mostly for the sake of understanding.

Important files: 
- [protos](https://github.com/cockroachdb/cockroach/pull/84646/files#diff-ffddfc79b1e0a1b701d33f9389168d399e6024c8e9b7e929cd622706914a8633)
- [interfaces](https://github.com/cockroachdb/cockroach/pull/84646/files#diff-3f82f6997d448fa3c06cf8941e3c8be6907f8abe098cc425b9df28117d113556)

### CODE REVIEW

**Review by ajwerner — COMMENTED** [review]
I hope my feedback makes sense. My hope is that we can keep the boundaries concept solely upstream of kvserver

File: pkg/keyvisualizer/key_visualizer.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+// │                        │       │                      │
+// └────────────────────────┘       └──────────────────────┘
+
+
+type BoundaryUpdate keyvispb.Boundaries
+
+// SpanStatsCollector manages the collection of store-level batch requests
+// per tenant. It opaquely manages the persistence of the last N tenant
+// histograms for fault-tolerance purposes.
+type SpanStatsCollector interface {
```
[the remark this answers is not in the document]
Comment (ajwerner) [reply pkg/keyvisualizer/key_visualizer.go:34]: I think this is the crux of the contention. I don't understand what you mean regarding coupling unless you're suggesting that you're going to get good sub-range statistics. I'm not sure how I'd solve that, but I think I'd solve it by augmenting the metadata that a range collects and stores, and not anything else. Sub-range sampling is hard, and, I assumed, out of scope.

[the remark this answers is not in the document]
Comment (ajwerner) [reply pkg/keyvisualizer/key_visualizer.go:34]: I have a hard time internalizing this data-flow as coupling. Or, put differently, yes, I do think the collector should be aware of the underlying concepts of the store (the collector depends on the kv layer) and not vice-versa.

**Comment by ajwerner** [comment]
@irfansharif and I sync'd up. I think the what he's proposing in terms of pushing down collection boundaries in order to get clear signal along stable boundaries which may align with sql concepts makes sense as an extension. It can cleanly fit into the design we're broadly talking about here. It also assumes that the tenant is collecting range-level data in order to evaluate and reformulate its boundary choices.

At this point, I think we ought to defer any boundary or state push-down into KV until later in this project. For now, the question at hand is how to store the compressed bucket data that has been collected. Whether it comes from compressing per-range data or from aggregated data which occurs along pre-defined boundaries (which themselves were derived from some combination of range data and sql data) isn't really relevant. For now, the simplest thing to do is continue on the path of processing a stream of range data into buckets data in the tenant, storing it in that tenant, and serving it to a UI.

**Review by andreimatei — COMMENTED** [review]
For me, an important thing is to have control over the rate of change in boundaries, and also over which boundaries change when we're forced to change some (I think we need to heavily prefer changes in buckets with low QPS). I think we need to imagine the storage of the samples as one timeseries per bucket (I hope that's what we'll have in the Obs Service), which makes understanding what happens when a bucket comes and goes more visceral.

So I'm not very sure about the bucket stability scheme, as I understand it: flip a coin for every sample and decide whether to keep all the buckets from the last sample (in which case the measurements in the current get "interpolated"), or start with the current ranges as boundaries and downsample them without consideration for the boundaries in the last sample. I'm not sure how to analyze the rate of change of this scheme; I think it's unpredictable (when you don't keep the boundaries from last sample, you might get a completely new set of buckets).

I think we should have a scheme where the the change in buckets is as controlled as possible. That's why I liked what I think the plan was at some point - to have a central controller that decides the boundaries, independent of the range splits. Why did we move away from that? Was it because we wanted to collect other measurements besides QPS for our buckets, which measurements cannot be easily decoupled from ranges? Like live bytes and garbage?

I think that indeed we should collect these other measurements. I wonder if the buckets on which we collect them should be the same as for QPS. I guess an argument for bundling all measurements together is that it doesn't seem too important for measurements other than QPS to be localized very precisely (e.g. if a huge range in terms of live bytes is adjacent to a bunch of small ranges, and we aggregate the size across them in a single bucket, I don't think anybody would care, as long as we don't cross index partitions). Whereas QPS is important to localize, so we might as well let the QPS measurements dictate the buckets for everything.

I'm thinking that, even if we start with ranges, and we collect data on every range without a possibility for stats boundaries within a range, we should still talk more about how the central collector would change the buckets over time.

Btw, besides being seemingly more clear how we control the buckets change, the other reason why I liked a controller is for distributing the work of bucketing the stats and reducing the amount of data that needs to be centralized. I think we should design a system that can handle a relatively large number of ranges without costing too much. You can imagine a serverless cluster with, say, 10k tenants, each with, say 1k ranges => 10M ranges => 30M replicas (or much more in MR). Collecting and processing one set of stats per range per sampling interval (say, with a 10s sampling interval) can add up to a core (which is probably the cap for a non-parallel implementation). Grantedly, this would be a big cluster where presumably we could find a way to isolate and account for this collector process, but still.

---

Since I understand this isn't shipping in 22.2, we can consider having the Obs Service deal with storing this data from the get-go. I haven't started working on timeseries storage, but I did start on event storage - so we could start ingesting the data as events.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84646#-:-N7rewAN67bUEumOcDm9:b-di78vi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @irfansharif, and @zachlite)

___
*[`pkg/keyvisualizer/key_visualizer.go` line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84646#-N7rIH-N40Kg9V6FJWLj:-N7rIH-O019Rj_29g4-a:bal8t83) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca72b47f8cd3ba4e9a9ef4f0e54a6673cdea5559/pkg/keyvisualizer/key_visualizer.go#L37)):*
> ```Go
> 	// SaveBoundaries will stash a tenant's desired collection boundaries to be
> 	// installed at the beginning of the next sample period.
> 	SaveBoundaries(boundaries keyvispb.Boundaries) error
> ```

`UpdateBoundaries` ?
___
*[`pkg/keyvisualizer/keyvispb/key_visualizer.proto` line 51 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84646#-N7s4KNJDq0bTOmY-hSz:-N7s4KNJDq0bTOmY-hT-:b-ldwpyf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca72b47f8cd3ba4e9a9ef4f0e54a6673cdea5559/pkg/keyvisualizer/keyvispb/key_visualizer.proto#L51)):*
> ```protobuf
> }
> 
> message CommonSpan {
> ```

what's "Common" for this type definition. Just `Span` seems more appropriate. But I hope we can get rid of it as I comment below.
___
*[`pkg/keyvisualizer/keyvispb/key_visualizer.proto` line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84646#-N7s4fb-0PEvzkmknE6u:-N7s4fb-0PEvzkmknE6v:b-ygpmlv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca72b47f8cd3ba4e9a9ef4f0e54a6673cdea5559/pkg/keyvisualizer/keyvispb/key_visualizer.proto#L54)):*
> ```protobuf
>   uint64 id = 1;
>   roachpb.Span span = 2;
> }
> ```

is this nullable?
___
*[`pkg/keyvisualizer/keyvispb/key_visualizer.proto` line 58 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84646#-N7s3tBH9a8juN7iYbkp:-N7s3tBH9a8juN7iYbkq:b9moad7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca72b47f8cd3ba4e9a9ef4f0e54a6673cdea5559/pkg/keyvisualizer/keyvispb/key_visualizer.proto#L58)):*
> ```protobuf
> message Bucket {
>   util.hlc.Timestamp timestamp = 1 [(gogoproto.nullable) = false];
>   uint64 span_id = 2;
> ```

what's a `span_id`?
If this is used only in `GetKeyVisSamplesResponse`to reference a `CommonSpan`, consider inlining both those messages in `GetKeyVisSamplesResponse`. And in fact, can we use indexes instead, and get rid of the `id` in `CommonSpan` (or get rid of `CommonSpan` altogether)?
___
*[`pkg/keyvisualizer/keyvispb/key_visualizer.proto` line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84646#-N7s3Y3H9G9sTd42Za5e:-N7s3Y3H9G9sTd42Za5f:bblp8fj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca72b47f8cd3ba4e9a9ef4f0e54a6673cdea5559/pkg/keyvisualizer/keyvispb/key_visualizer.proto#L59)):*
> ```protobuf
>   util.hlc.Timestamp timestamp = 1 [(gogoproto.nullable) = false];
>   uint64 span_id = 2;
>   uint64 batch_requests = 3;
> ```

I'd vote to divorce a little bit from `BatchRequest` RPCs and talk only about `Requests`.
In fact, I thought we already switched to counting individual requests in `BatchRequest` for the purpose of the QPS stats... Maybe we didn't.
___
*[`pkg/keyvisualizer/keyvispb/key_visualizer.proto` line 95 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84646#-N7rJNKcEp_pr3mIUJVK:-N7rJNKcEp_pr3mIUJVL:b-qwge8k) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca72b47f8cd3ba4e9a9ef4f0e54a6673cdea5559/pkg/keyvisualizer/keyvispb/key_visualizer.proto#L95)):*
> ```protobuf
>   // GetKeyVisSamples is issued by the Browser. It returns a space-efficient
>   // representation of samples collected between `start` and `end`.
>   rpc GetKeyVisSamples(GetKeyVisSamplesRequest) returns
> ```

Are `KeyVisSamples` here the same as "samples" everywhere else? If so, make the name uniform in these RPCs.

**Review by irfansharif — COMMENTED** [review]
Meta: I'm aware of at least three separate 1:1s that have happened discussing these interfaces by the participants above. Across these 1:1s and discussion above, I've heard both (i)  the desire to trim this down to fit 22.2 in three weeks, and (ii) this not shipping in 22.2. Some of the 22.2 scheduling constraints have led to suggestions to focus entirely on the "job and above", by letting KV only emit range-level stats for now. Getting into a room together could help (if anything to decide what implementation order we ought to pursue). Below I'll try to summarize the final state of what I currently think we should get to design wise, ignoring all scheduling constraints. This has not changed from what was spelled out in this internal gdoc: https://docs.google.com/document/d/1Cp-JVh460ezFaxd0ZglVFGa2orUvWPNXPJKpOgEfV6s/edit.

A. Per-tenant controller explicitly telling KV the exact key boundaries to collect statistics over. This controller is free to use whatever data is needed to decide these boundaries:
   - Previously installed boundaries, if any;
   - Data from SQL, how specific tables/indexes/partitions are encoded, in order to collect data for SQL constructs;
   - Data from KV, what the range boundaries are, within a range what key segments are observing load (obtained perhaps from the reservoir samples maintained for range splits);
   - Anything else in the future we deem useful.

Whether collection boundaries are kept stable over time, where we put more "collection buckets" to use, etc. are implementation details hidden behind an interface that effectively spits out a list of `[]roachpb.Span`. This scheme has all the benefits Andrei spells out above, and a few others:
   - Collection data is not proportional to # of replicas;
   - When combining results across stores, we don't have to worry about range splits/merges happening while we're talking to each store one-by-one (for a range [a,c) that was split into [a,b) and [b,c), how does one combine the results from a store pre-split with the results from a store post-split?)
   - Allows more direct control over boundary stability, and consequently space efficiency for B below;
   - Obviates the need for any sort of interpolation. If you're looking to present stats over `[a,b)`, corresponding to an index perhaps, and only have collections over `[a1,b1)`, how does one interpolate from one to the other? (this doesn't make sense to me semantically -- if interpolating by # of keys or bytes, what does it mean when interpolating from [a,b) to [a,b) where the number of keys have changed dramatically between the collections?)
   - Does not couple the visual artifact to range boundaries KV's introducing for whatever reason (zone configs, load, etc.). This lets the the visualizer adapt to both small (tens of ranges) and large (thousands of ranges) clusters. It makes it viable for serverless users. It sidesteps the questions of "sub-range statistics".
   - Key span statistics aren't tied to range lifetimes. Ranges could split/merge/get resuscitated through loss-of-quorum tooling and we'd still maintain span statistics since these counters aren't maintained on Replica objects. Without this, it’s possible to have statistics that correspond to the replica's lifetime and not "since the last collection".

We can defer some of this work until later, to accommodate whatever release schedule we're aiming for, but this structure feels important. A simple implementation of this interface could be what Zach prototyped -- set collection boundaries every 15m to whatever the range boundaries happen to be at that point in time. This could be the first version we ship.

B. Per-tenant component that periodically retrieves span statistics from KV and persists it in a format that's:
   - Space-efficient (if stored in SQL tables, perhaps pulling bucket metadata into it's own tables; some other "one timeseries per bucket" form if storing in the Obs Service).
   - Access-efficient for the visualizer itself (are we likely to query by timestamps? Or by keyspans? Both? What's the payload we're looking to transmit?)

This component should really only merge adjacent collection bucket data to fit whatever budget it needs to stay under ("1000 data points every time window"). What boundaries it decides to use (and how stable they are over time) could sit behind an interface. We do no "interpolation" here, if we want to render stats from [a,e), the controller described above must ensure we're collecting over `[a,_), [_,e)`.

C. Per-store span stats collector that listens in on whatever key boundaries it needs to collect statistics over, for each tenant, and maintains counters by intercepting all batch requests/responses to the store. If a batch request straddles multiple collection buckets, we'll increment all of them. If we were observing a scan with a limit, and didn't in fact touch the entire span in the batch request, we'll know as part of the batch response and adjust counters accordingly.

---

After this plumbing in place, our first pass can be something that just collects stats over range boundaries (though having those boundaries installed explicitly through A, and acted upon by C). Once that sees active use, it'll give us better intuitions for the big question we sidestepped: how the {collection,visualization} boundaries evolve over time? This has many facets, including alignment with SQL indexes/partitions, stability over time, shifting resolution to keyspans of interest away from non-interesting ones. The structure above should however let us prototype various ideas under a well defined interface, and in future releases in a backwards compatible way. If the visualizer simply followed range boundaries, that's only one of many possible implementations of this interface. If we explore visualizing something other than '# of batch requests', this structure lets us get there too (CPU time/disk IOPs/bandwidth per-span come to mind).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84646#-:-N7sA-jM1oFLjCpAQUp3:b-lvnu6d)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, and @zachlite)

**Pull request closed without merging**